src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -63,68 +72,229 @@ void BatchNorm<InputDataType, OutputDataType>::Reset()
-  if (deterministic)
-  {
-    // Normalize the input and scale and shift the output.
-    output = input.each_col() - runningMean;
-    output.each_col() %= gamma / arma::sqrt(runningVariance / count + eps);
-    output.each_col() += beta;
-  }
-  else
+  const size_t batchSize = input.n_cols;
+
+  // Set size of output equal to the size of input.
+  output.set_size(arma::size(input));
+
+  // We will calculate mini batch norm on each channel / feature map.
+  if (!deterministic)
   {
-    mean = arma::mean(input, 1);
-    variance = arma::var(input, 1, 1);
+    Log::Assert(input.n_cols > 0, "Empty input. Batch size must be \
+        greater than zero.");
 
-    // Normalize the input.
-    output = input.each_col() - mean;
-    inputMean = output;
-    output.each_col() /= arma::sqrt(variance + eps);
+    Log::Assert(input.n_rows % size == 0, "Input features must be divisible \
+        by feature maps.");
 
-    // Use Welford method to compute the sample variance and mean.
-    for (size_t i = 0; i < input.n_cols; i++)
+    // Input size refers to inputWidth * inputHeight.
+    // If inputSize equals one, input corresponds to output from linear layer.
+    const size_t inputSize = input.n_rows / size;
+
+    // Linear output.
+    if (inputSize == 1)
     {
-      count += 1;
+      mean = arma::mean(input, 1);
+      variance = arma::var(input, 1, 1);
+
+      // Normalize the input.
+      output = input.each_col() - mean;
+      inputMean = output;
+      output.each_col() /= arma::sqrt(variance + eps);
+
+      // Use Welford method to compute the sample variance and mean.
+      for (size_t i = 0; i < input.n_cols; i++)
+      {
+        count += 1;
+
+        OutputDataType diff = input.col(i) - runningMean;
+        if (average)
+        {
+          runningMean = runningMean + diff / count;
+          runningVariance += diff % (input.col(i) - runningMean);
+        }
+        else
+        {
+          runningMean = (1 - momentum) * runningMean + momentum * mean;
+          runningVariance = (1 - momentum) * runningVariance +
+              momentum * variance;
+        }
+      }
+
+      // Reused in the backward and gradient step.
+      normalized = output;
 
-      OutputDataType diff = input.col(i) - runningMean;
-      runningMean = runningMean + diff / count;
-      runningVariance += diff % (input.col(i) - runningMean);
+      // Scale and shift the output.
+      output.each_col() %= gamma;
+      output.each_col() += beta;
+      return;
     }
 
-    // Reused in the backward and gradient step.
-    normalized = output;
+    // Calculate mean and variance over all channels.
+    mean = OutputDataType(1, size, arma::fill::zeros);
+    variance = OutputDataType(1, size, arma::fill::zeros);
+
+    // Initialize output to same size and values for convience.
+    output = input;
+    normalized.set_size(arma::size(input));
+    inputMean.set_size(arma::size(input));
+    double nElements = input.n_elem / size;
```

Comment:
```suggestion
    double nElements = (double) input.n_elem / size;
```

Just to keep the static code analysis job green.

Reply:
Sure, Will make the changes.